clang/lib/Analysis/CFG.cpp

```diff
@@ -6298,10 +6308,60 @@ static bool isImmediateSinkBlock(const CFGBlock *Blk) {
   // at least for now, but once we have better support for exceptions,
   // we'd need to carefully handle the case when the throw is being
   // immediately caught.
-  if (llvm::any_of(*Blk, [](const CFGElement &Elm) {
+  if (llvm::any_of(*Blk, [](const CFGElement &Elm) -> bool {
+        if (std::optional<CFGStmt> StmtElm = Elm.getAs<CFGStmt>())
+          return isa<CXXThrowExpr>(StmtElm->getStmt());
+        return false;
+      }))
+    return true;
+
+  auto HasNoReturnCall = [&](const CallExpr *CE) {
+    if (!CE)
+      return false;
+
+    auto *FD = CE->getDirectCallee();
+
+    if (!FD)
+      return false;
+
+    auto *CanCD = FD->getCanonicalDecl();
+    auto *DefFD = CanCD->getDefinition();
+    auto NoRetAttrOpt = CanCD->getAnalyzerNoReturn();
+    auto NoReturn = false;
+
+    if (!NoRetAttrOpt && DefFD && DefFD->getBody()) {
+      // HACK: we are gonna cache analysis result as implicit
+      // `analyzer_noreturn` attribute
+      auto *MutCD = const_cast<FunctionDecl *>(CanCD);
```

Comment:
I see now how you use the implicit attribute to cache this deduced noreturn property.
I agree that this is a hack. It's also unclear to me if this recursively descend approach would work with recursive calls. For example:
```c++
int fib(int n) {
  return fib(n - 2) + fib(n - 1);
}
```

How I understand, while building the CFG of `fib`, we would see a `CallExpr` to `fib`, thus recurse, and then again and again - before it could attach the implicit attribute.
I wish you could demonstrate this in a test.

There are ways around this, but it's not easy during CFG construction.
I'll think about this problem and come back if I have any ideas.